src/Time/History.hpp

```diff
@@ -3,400 +3,910 @@
+  /// returns a class derived from `ConstUntypedHistory<UntypedVars>`.
+  /// Any modifications to this class invalidate the returned object.
+  ConstUntypedAccess untyped() const { return ConstUntypedAccess(*this); }
+
+  /// Mutable, type-erased access to the history.  This method returns
+  /// a class derived from `MutableUntypedHistory<UntypedVars>`.  Any
+  /// modifications to this class invalidate the returned object,
+  /// except for those performed through the object itself.
+  MutableUntypedAccess untyped() { return MutableUntypedAccess(this); }
+
+  /// Get or set the order the time stepper is running at.  Many time
+  /// steppers expect this to have a particular value.  This has no
+  /// effect on the storage of past data.
+  /// @{
+  size_t integration_order() const { return integration_order_; }
+  void integration_order(const size_t new_integration_order) {
+    integration_order_ = new_integration_order;
+  }
+  /// @}
 
-  HistoryIterator(const UntypedHistory<T>* const history,
-                  const difference_type offset);
+  /// Type and value used to indicate that a record is to be created
+  /// without the `value` field set.
+  /// @{
+  struct NoValue {};
+  static constexpr NoValue no_value{};
+  /// @}
 
-  const UntypedHistory<T>* history_;
-  difference_type offset_;
-};
+  /// Insert a new entry into the history.  It will be inserted as a
+  /// step or substep as appropriate.
+  ///
+  /// The supplied `time_step_id` must be newer than the current
+  /// newest entry, and if the substep of `time_step_id` is nonzero,
+  /// the id must be consistent with the existing data for the current
+  /// step.
+  ///
+  /// If `no_value` is passed, the created record will not have its
+  /// `value` field set.  This is primarily useful for histories for
+  /// the implicit portions of IMEX time steppers.
+  ///
+  /// The `value` and `derivative` data will be copied into cached
+  /// allocations, if any are available.
+  /// @{
+  void insert(const TimeStepId& time_step_id, const Vars& value,
+              const DerivVars& derivative);
+  void insert(const TimeStepId& time_step_id, NoValue /*unused*/,
+              const DerivVars& derivative);
+  /// @}
 
-/// \ingroup TimeSteppersGroup
-/// Iterator over the derivatives stored in a History object.
-///
-/// This returns typed data, and so cannot be obtained from an
-/// UntypedHistory.  Type-erased derivatives can be obtained from
-/// HistoryIterator::derivative().
-template <typename DerivVars>
-class DerivIterator {
-  using Base =
-      typename std::deque<std::tuple<TimeStepId, DerivVars>>::const_iterator;
+  /// Insert data at the start of the history.  This can be useful
+  /// when initializing a multistep integrator.
+  ///
+  /// This should only be used for initialization, and cannot be used
+  /// for substep data.  The supplied `time_step_id` must be older
+  /// than the current first entry.
```

Comment:
Added a comparison to `push_front`, since that is really what this is.  "Older" has its usual physics meaning of "from an earlier time" (unless time is running backwards).  I changed to "earlier", which can't be confused with any sort of C++ or container-related thing.  (I usually use "before" for this relation, but that's not great when discussing an ordered container.)